src/artifact-uploader.cpp

```diff
@@ -19,28 +27,36 @@ namespace connect
 class ArtifactUploader::Impl
 {
 public:
+    Impl()
+        : accountId_(-1)
+        , cameraId_(-1)
+        , done_(false)
+        , timeoutToCompleteUploadSec_(0)
+    {
+    }
+
     ~Impl();
 
     Status init(const ArtifactUploader::Configuration& cfg,
                 ArtifactUploader::ClientConfigCallback* configCallback);
 
-    // all upload* methods are asynchronous, non-blocking, take ownership
-    // of data passed to them
-    Status uploadBackground(const timestamp_t& timestamp, PayloadHolderPtr payload);
-    Status uploadObjectStream(const ObjectStream& stream, PayloadHolderPtr payload);
-    Status uploadFlipbook(const Flipbook& flipbook, PayloadHolderPtr payload);
-    Status uploadEvent(const timestamp_t& timestamp, move_ref<Events> data);
-    Status uploadCount(move_ref<Counts> counts);
-
 private:
-    OutputControllerPtr outputController_;
-    UploadTaskQueuerPtr uploadTaskQueuer_;
+    friend class ArtifactUploader;
+
+    void threadFunc();
+
+    unique_ptr<Client>::t client_;
+    id_t accountId_;
+    id_t cameraId_;
+    UploadQueuePtr queue_;
+    boost::thread thread_;
+    volatile bool done_;
```

Comment:
Is "volatile" sufficient for the purpose?
See [here](http://stackoverflow.com/questions/4557979/when-to-use-volatile-with-multi-threading), [here](http://stackoverflow.com/questions/2484980/why-is-volatile-not-considered-useful-in-multithreaded-c-or-c-programming) and [here](http://www.codingstandard.com/rule/18-2-3-do-not-share-volatile-data-between-threads/)